pkg/api/query/grpc.go

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package v1
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/prometheus/promql"
+	"github.com/thanos-io/thanos/pkg/api/query/querypb"
+	"github.com/thanos-io/thanos/pkg/query"
+	"github.com/thanos-io/thanos/pkg/store/labelpb"
+	"github.com/thanos-io/thanos/pkg/store/storepb/prompb"
+	"google.golang.org/grpc"
+)
+
+type GRPCAPI struct {
+	now                         func() time.Time
+	queryableCreate             query.QueryableCreator
+	queryEngine                 func(int64) *promql.Engine
+	defaultMaxResolutionSeconds time.Duration
+}
+
+func NewGRPCAPI(now func() time.Time, creator query.QueryableCreator, queryEngine func(int64) *promql.Engine, defaultMaxResolutionSeconds time.Duration) *GRPCAPI {
+	return &GRPCAPI{
+		now:                         now,
+		queryableCreate:             creator,
+		queryEngine:                 queryEngine,
+		defaultMaxResolutionSeconds: defaultMaxResolutionSeconds,
+	}
+}
+
+func RegisterQueryServer(queryServer querypb.QueryServer) func(*grpc.Server) {
+	return func(s *grpc.Server) {
+		querypb.RegisterQueryServer(s, queryServer)
+	}
+}
+
+func (grpcAPI *GRPCAPI) Query(request *querypb.QueryRequest, server querypb.Query_QueryServer) error {
```

Comment:
```suggestion
func (g *GRPCAPI) Query(request *querypb.QueryRequest, server querypb.Query_QueryServer) error {
```

nit: We can make it shorter if we want. (not a blocker)